catalog/pages/components/table-view.md

```diff
@@ -0,0 +1,24 @@
+``` react
+state: {
+    filters: {}
+}
+---
+<TableView
+    rows={data}
+    columns={[
+        {label: 'Title', dataKey: 'title' },
+        {label: 'Genre', dataKey: 'genre' },
+        {label: 'Rating from 10', dataKey: 'rating'}
+    ]}
+    filters={[
+        {label: 'Genre', dataKey: 'genre', value: state.filters['genre']},
+        {label: 'Rating', dataKey: 'rating', value: state.filters['rating']}
+    ]}
+    onFilterChange={(dataKey, value) => setState({
```

Comment:
I made the decision in the context and scale/scope of the challenge.

I followed the idiom of the existing code - stateless components. Main inspiration was `select.jsx` and `select.md`. This decision required to move the state management - filter control - to parent component.

Of course, a stateless component improves testability (admittedly while moving the complexity to its parent).

For `TableView` to be able to control filters, we need it to be stateful. In this case, the parent should only dictate defaultValue of a filter. Then the API could roughly be:

```
<TableView rows columns
    filters=[{dataKey: 'genre', label: 'Genre', defaultValue: ['drama']}]
```
whereby the TableView would set its initial filter-state in constructor by using `props.filters.defaultValue`. Roughly,

```
constructor(props) {
    super(props)
    this.state = sanitizeFilters(props.filters)
}
```
The above code incidentally also resolves the [next issue](https://github.com/jayesh-bhoot/challenge-table-view/pull/1#discussion_r309816712), as now sanitization for `undefined` happens only once in constructor.


Of course, this is when we assume that the parent dictates which columns become filters and that the number of filtrable-columns wont change (say: a filtrable-columns added or removed).

In such a broadened usecase (number of filtrable-columns can change), I would not pass filters at all as props, or at best, only put in `defaultFilters`. Roughly,

```
<TableView rows columns
    defaultFilters=[{dataKey, label, defaultValue}]
```

If we don't pass the `defaultFilters`, then the parent cannot dictate the filters at all, and the whole onus of default filtrable-columns and `defaultValue`s of all filters falls on `TableView`.

In a further larger usecase, if I had to persist the filtered view across reload, then I would have put the state in redux store.

Reply:
I wouldn't say that `Select` is a stateless component as stateless components uses functions not classes, anyway it wasn't said or intended to make any of the created components stateless, that decision was left up to you.
I appreciate you explaining how to implement more features for the component but this is a complexity I prefer to avoid now.

Reply:
> I wouldn't say that Select is a stateless component as stateless components uses functions not classes
I used the term liberally, to mean that we dont hold state in the component. My bad.

> I appreciate you explaining how to implement more features for the component but this is a complexity I prefer to avoid now.
Agreed. I was trying to put my line of thought into words to explain how I would have changed the design depending on context.

Anyhow, you are the judge.